cpp/src/parquet/column_writer.cc

```diff
@@ -1364,6 +1367,41 @@ class TypedColumnWriterImpl : public ColumnWriterImpl, public TypedColumnWriter<
                          int64_t num_levels, const ::arrow::Array& array,
                          ArrowWriteContext* context, bool maybe_parent_nulls);
 
+  template <typename ArrowType>
+  Status WriteArrowSerialize(const int16_t* def_levels, const int16_t* rep_levels,
+                             int64_t num_levels, const ::arrow::Array& array,
+                             ArrowWriteContext* ctx, bool maybe_parent_nulls);
+
+  Status WriteArrowZeroCopy(const int16_t* def_levels, const int16_t* rep_levels,
+                            int64_t num_levels, const ::arrow::Array& array,
+                            ArrowWriteContext* ctx, bool maybe_parent_nulls) {
+    const auto& data = static_cast<const ::arrow::PrimitiveArray&>(array);
+    const T* values = nullptr;
+    // The values buffer may be null if the array is empty (ARROW-2744)
+    if (data.values() != nullptr) {
+      values = reinterpret_cast<const T*>(data.values()->data()) + data.offset();
+    } else {
+      DCHECK_EQ(data.length(), 0);
+    }
+    bool no_nulls =
+        this->descr()->schema_node()->is_required() || (array.null_count() == 0);
+
+    if (!maybe_parent_nulls && no_nulls) {
+      PARQUET_CATCH_NOT_OK(WriteBatch(num_levels, def_levels, rep_levels, values));
```

Comment:
What if `array.null_count() != 0` here?

Reply:
This is the existing logic, I haven't changed it. 
I there are actually null values in the array not just being nullable, then we need to pick the slower path.

Reply:
> What if array.null_count() != 0 here?

It should not happen when the field is required but has null values: https://github.com/apache/arrow/blob/main/cpp/src/parquet/column_writer.cc#L1324

Reply:
Ok, thanks. Ideally we would have a `DCHECK` but since this is just moving code around we'll live without it.